korim/src/commonMain/kotlin/com/soywiz/korim/bitmap/BitmapSlice.kt

```diff
@@ -281,68 +329,313 @@ abstract class BmpSlice(
     val right get() = bounds.right
     val bottom get() = bounds.bottom
 
+    var bmpCoords: BmpCoordsWithT<ISizeInt>
+        private set
+
+    init {
+        if (bounds.x < 0) {
+            bounds.width += bounds.x
+            bounds.x = 0
+        }
+        if (bounds.y < 0) {
+            bounds.height += bounds.y
+            bounds.y = 0
+        }
+        if (bounds.width < 0) bounds.width = 0
+        if (bounds.height < 0) bounds.height = 0
+        if (bmpCoords != null) {
+            this.bmpCoords = bmpCoords.withImageOrientation(imageOrientation)
+        } else {
+            val tlX: Float
+            val tlY: Float
+            val trX: Float
+            val trY: Float
+            val brX: Float
+            val brY: Float
+            val blX: Float
+            val blY: Float
+
+            if (parentCoords != null) {
+                // Calculate bmpCoords based on parentCoords
+                val dx = parentCoords.br_x - parentCoords.tl_x
+                val dy = parentCoords.br_y - parentCoords.tl_y
+                val x = bounds.x.toFloat()
+                val y = bounds.y.toFloat()
+                val w = bounds.width.toFloat()
+                val h = bounds.height.toFloat()
+                val bw = parentCoords.width.toFloat()
+                val bh = parentCoords.height.toFloat()
+
+                if (parentCoords.tl_x == parentCoords.tr_x) {
+                    // Base is rotated
+                    tlX = parentCoords.tl_x + y / bh * dx
+                    tlY = parentCoords.tl_y + x / bw * dy
+                    when (imageOrientation.rotation) {
+                        ImageOrientation.Rotation.R90, ImageOrientation.Rotation.R270 -> {
+                            brX = tlX + w / bh * dx
+                            brY = tlY + h / bw * dy
+                        }
+                        else -> {
+                            brX = tlX + h / bh * dx
+                            brY = tlY + w / bw * dy
+                        }
+                    }
```

Comment:
Looking at this, can we simplify this kind of things?

For example:

```kotlin
val rotated = imageOrientation.rotation == ImageOrientation.Rotation.R90 || imageOrientation.rotation == ImageOrientation.Rotation.R270
val xDim = if (rotated) w else h
val yDim = if (rotated) w else h
brX = tlX + xDim / bh * dx
brY = tlY + yDim / bw * dy
```

The idea is to reduce duplicated arithmetic and ifs as much as possible so the code is as straightforward as possible. Since there are many big ifs and stuff there is a lot of code and it is hard to view all the edge cases, and changing something might lead to not changing all the paths, etc.
Can you try to figure out patterns in the code and simplify this code a bit more?